### PR TITLE
[CCI] Add bluebird replaces for src/dev

### LINKED ISSUE #3662: Eliminate Bluebird in favor of native async functionality
Bluebird is a open-source JavaScript library for working with asynchronous code. It is a Promise library that provides a number of useful features and enhancements to the built-in JavaScript Promise object, including performance optimizations and additional methods.

The task is to replace all instances of Bluebird with native async functionality in the codebase. This should be done using the async and await keywords, which have been natively available for over 5 years in all relevant environments. Promises have also been available for longer.

- [ ] Identify all instances where Bluebird is used in the codebase
- [ ] Replace the Bluebird calls with the equivalent native async functionality using async and await
   - [ ] Below are the list of bluebird functions that are currently used and their equivalent native implementation
```
         all -> Promise.all
         settle -> Promise.allSettled
         resolve -> Promise.resolve
         reject-> Promise.reject
         setSchedular -> queueMicrotask
         fromNode ->  const util = require('util');
                                await util.promisify(readFileCallback);
         delay ->  function delay(ms) {
                        return new Promise(resolve => setTimeout(resolve, ms));}
         mapAsync -> function mapAsync(array, mapper) {
                               return Promise.all(array.map(mapper));}
         promisify -> await util.promisify(fs.readFile);
```
- [ ] Test the updated code to ensure that it is functioning as expected

### PR DESCRIPTION
**Pull request by Nicksqain**
### Issues Resolved

- closes #3662
### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/dev/jest/setup/polyfills.js [hunk 1]
```diff
@@ -30,9 +30,8 @@
 
 // bluebird < v3.3.5 does not work with MutationObserver polyfill
 // when MutationObserver exists, bluebird avoids using node's builtin async schedulers
-const bluebird = require('bluebird');
-bluebird.Promise.setScheduler(function (fn) {
-  global.setImmediate.call(global, fn);
+queueMicrotask(() => {
```
Comment (Nicksqain) [inline src/dev/jest/setup/polyfills.js:33]: I haven't quite figured it out, so I'm not sure what the right solution is, can you tell me what the right solution should be?

File: src/dev/mocha/__tests__/junit_report_generation.js [hunk 2]
```diff
@@ -62,7 +62,7 @@ describe('dev/mocha/junit report generation', () => {
 
     mocha.addFile(resolve(PROJECT_DIR, 'test.js'));
     await new Promise((resolve) => mocha.run(resolve));
-    const report = await fcb((cb) => parseString(readFileSync(XML_PATH), cb));
+    const report = await promisify(parseString)(readFileSync(XML_PATH));
```
Comment (AMoo-Miki) [inline src/dev/mocha/__tests__/junit_report_generation.js:65]: We should just use `readFile` from `fs/promises` and `parseStringPromise` from `xml2js`.

```suggestion
    const report = await parseStringPromise(await readFile(XML_PATH));
```

File: src/dev/notice/bundled_notices.js [hunk 3]
```diff
@@ -31,16 +31,16 @@
...
 
 export async function getBundledNotices(packageDirectory) {
   const pattern = resolve(packageDirectory, '*{LICENSE,NOTICE}*');
-  const paths = await fcb((cb) => glob(pattern, cb));
+  const paths = await promisify(glob)(pattern);
   return Promise.all(
     paths.map(async (path) => ({
       path,
-      text: await fcb((cb) => readFile(path, 'utf8', cb)),
+      text: await promisify(readFile)(path, 'utf8'),
```
Comment (AMoo-Miki) [inline src/dev/notice/bundled_notices.js:43]: We should use `fs/promises`

File: src/dev/precommit_hook/get_files_for_commit.js [hunk 4]
```diff
@@ -44,7 +44,7 @@ import { File } from '../file';
 export async function getFilesForCommit() {
   const simpleGit = new SimpleGit(REPO_ROOT);
 
-  const output = await fcb((cb) => simpleGit.diff(['--name-status', '--cached'], cb));
+  const output = await promisify(simpleGit.diff)(['--name-status', '--cached']);
```
Comment (AMoo-Miki) [inline src/dev/precommit_hook/get_files_for_commit.js:47]: According to the docs, we should be able to do this without `promisify`, like this:
```suggestion
  const output = await simpleGit.diff(['--name-status', '--cached']);
```

**Review by AMoo-Miki — CHANGES_REQUESTED** [review]

File: src/dev/jest/setup/polyfills.js (hunk 1 above)
Comment (ananzh) [reply src/dev/jest/setup/polyfills.js:33]: I think you could just removing Bluebird here. The scheduler will no longer be set to use setImmediate, and will use the default native Promise scheduling instead. Unless the code depends on the specific scheduling behavior of Bluebird, this should not be a problem.

```
const MutationObserver = require('mutation-observer');
Object.defineProperty(window, 'MutationObserver', { value: MutationObserver });

require('whatwg-fetch');
```

Comment (AMoo-Miki) [reply src/dev/jest/setup/polyfills.js:33]: Anan is right; being a Jest polyfill, we will know if we have a problem instantly due to test failures.

**Comment by abbyhu2000** [comment]
Convert to draft for now, can be marked as ready once the comments are addressed.

**Comment by ananzh** [comment]
@Nicksqain Due to an extended period without updates, we're going to close this issue for now. We appreciate your contribution and understand that priorities and availability can change. Please feel free to reopen this issue when you have the opportunity to revisit it. Thank you for your understanding.

**Pull request closed without merging**